File: c++/leetcode/451. Sort Characters By Frequency.cpp
class Solution
{
public:
  string frequencySort(string s)
  {
    unordered_map<char, int> characters;
    priority_queue<pair<int, char>> pq;

    string ans = "";

    for (auto i : s)
      characters[i]++;

    for (auto i : characters)
    {
      pq.push(make_pair(i.second, i.first));
    }

    while (pq.size())
    {
      auto tp = pq.top();
      pq.pop();
      int times = tp.first;
      while (times--)
        ans += tp.second;
    }

    return ans;
  }
};

File: javascript/leetcode/451. Sort Characters By Frequency.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function (s) {
  let ans = "";
  let seen = new Map();

  for (let i of s) seen.has(i) ? seen.set(i, seen.get(i) + 1) : seen.set(i, 1);

  const pq = Array.from(seen.entries());

  pq.sort((a, b) => b[1] - a[1]);

  pq.forEach((el) => {
    let times = el[1];
    while (times--) ans += el[0];
  });

  return ans;
};
